flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoTableSourceScanRule.java

```diff
@@ -52,208 +54,324 @@
         return DynamicSourceUtils.isUpsertSource(table.catalogTable(), table.tableSource())
                 || DynamicSourceUtils.isSourceChangeEventsDuplicate(
                         table.catalogTable(), table.tableSource(), config);
     }
 
-    /**
-     * Push the used physical column and metadata into table source. The returned value is used to
-     * build new table schema.
-     */
-    private static RowType applyPhysicalAndMetadataPushDown(
-            NestedSchema nestedSchema,
-            List<String> metadataKeys,
-            RowType originType,
-            List<SourceAbilitySpec> sourceAbilitySpecs) {
-        // TODO: supports nested projection for metadata
-        List<NestedColumn> usedMetaDataFields = new LinkedList<>();
-        int physicalCount = originType.getFieldCount() - metadataKeys.size();
-        List<String> fieldNames = originType.getFieldNames();
-
-        // rm metadata in the tree
-        for (int i = 0; i < metadataKeys.size(); i++) {
-            NestedColumn usedMetadata =
-                    nestedSchema.columns().remove(fieldNames.get(i + physicalCount));
-            if (usedMetadata != null) {
-                usedMetaDataFields.add(usedMetadata);
-            }
+    private List<RexNode> getPrimaryKeyProjections(LogicalTableScan scan) {
+        final TableSourceTable source = scan.getTable().unwrap(TableSourceTable.class);
+        final ResolvedSchema schema = source.catalogTable().getResolvedSchema();
+        if (!schema.getPrimaryKey().isPresent()) {
+            return Collections.emptyList();
+        }
+
+        final FlinkTypeFactory typeFactory = unwrapTypeFactory(scan);
+        final UniqueConstraint primaryKey = schema.getPrimaryKey().get();
+        return primaryKey.getColumns().stream()
+                .map(
+                        columnName -> {
+                            final int idx = scan.getRowType().getFieldNames().indexOf(columnName);
+                            final Column column =
+                                    schema.getColumn(idx)
+                                            .orElseThrow(
+                                                    () ->
+                                                            new TableException(
+                                                                    String.format(
+                                                                            "Column at index %d not found.",
+                                                                            idx)));
+                            return new RexInputRef(
+                                    idx,
+                                    typeFactory.createFieldTypeFromLogicalType(
+                                            column.getDataType().getLogicalType()));
+                        })
+                .collect(Collectors.toList());
+    }
+
+    private RowType performPushDown(
```

Comment:
I found it difficult to split this method up further without being able to have state (maybe a short-coming of the singleton pattern for rules) and while still keeping it readable. However, I do think there are nice chunks of separated logic now.